File: mostly asked/Twosum.java
import java.util.HashMap;

public class Twosum {
    public static int[] twoSumloops(int n, int []arr, int target) {
        int[] ans = new int[2];
        ans[0] = ans[1] = -1;
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (arr[i] + arr[j] == target) {
                    ans[0] = i;
                    ans[1] = j;
                    return ans;
                }
            }
        }
        return ans;

         
    }

    public static int[] twosumtwopointer(int[] arr, int target) {
        int left = 0;
        int right = arr.length - 1;
        while (left < right) {
            int sum = arr[left] + arr[right];
            if (sum == target) {
                return new int[]{left, right};
            } else if (sum < target) {
                left++;
            } else {
                right--;
            }
        }
        return new int[]{-1, -1};
    }

    public static int[] twosumhashmap(int[] arr, int target) {
        HashMap<Integer, Integer> hashmap = new java.util.HashMap<>();
        for (int i = 0; i < arr.length; i++) {
            int complement = target - arr[i];
            if (hashmap.containsKey(complement)) {
                return new int[]{hashmap.get(complement), i};
            }
            hashmap.put(arr[i], i);
        }
        return new int[]{-1, -1};
    }

    public static void main(String args[]) {
        int n = 5;
        int[] arr = {2, 6, 5, 8, 11};
        int target = 17;
        int[] ans = twoSumloops(n, arr, target);
        System.out.println("This is the answer for variant 2: [" + ans[0] + ", "
                           + ans[1] + "]");

        ans=twosumtwopointer(arr,target);
        System.out.println("This is the answer for variant 3: [" + ans[0] + ", "
                           + ans[1] + "]");
        ans=twosumhashmap(arr,target);
        System.out.println("This is the answer for variant 4: [" + ans[0] + ", "
                           + ans[1] + "]");
    }

}

File: Leetcode/mostly asked/Twosum.py
# method 1  O(n^2)

def Twosumloops(arr,target):
    i=0
    for i in range(len(arr)):
        for j in range(i+1,len(arr)):
            if arr[j]+arr[i]==target:
                return [i,j]
            
    return [-1,-1]

# method 2 O(nlogn) two pointer approach

def Twosumtwopointer(arr,target):
    
    left=0
    right=len(arr)-1
    while left < right:
        if arr[left] + arr[right] == target:
            return [left,right]
        elif arr[left] + arr[right] < target:
            left+=1
        else:
            right-=1
    return [-1,-1]

# method 3 O(n) using hashmap

def Twosumhashmap(arr,target):
    hashmap={}  # store numbers seen so far with index as value
    for i in range(len(arr)):
        complement=target-arr[i]
        if complement in hashmap:
            return [hashmap[complement],i]
        hashmap[arr[i]]=i
    return [-1,-1]

if __name__ == "__main__":
    n=5
    arr=[2,3,4,5]
    target=9
    ans=Twosumhashmap(arr,target)
    print("This is the answer for variant 3:", ans)
    ans2=Twosumtwopointer(arr,target)
    print("This is the answer for variant 2:", ans2)
    ans3=Twosumloops(arr,target)
    print("This is the answer for variant 1:", ans3)
    
    